libsweep/src/win/serial.cc

```diff
@@ -0,0 +1,351 @@
+#include "serial.h"
+
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <cstdio>
+#include <string>
+
+#include <windows.h>
+
+namespace sweep {
+namespace serial {
+
+typedef struct error {
+  const char* what; // always literal, do not free
+} error;
+
+typedef struct device {
+  HANDLE h_comm;
+  OVERLAPPED os_reader;
+  BOOL waiting_on_read;      // Used to prevent creation of new read operation if one is outstanding
+  DWORD read_timeout_millis; // timeout interval for entire read operation
+} device;
+
+// Constructor hidden from users
+static error_s error_construct(const char* what) {
+  SWEEP_ASSERT(what);
+
+  auto out = new error{what};
+  return out;
+}
+
+const char* error_message(error_s error) {
+  SWEEP_ASSERT(error);
+
+  return error->what;
+}
+
+void error_destruct(error_s error) {
+  SWEEP_ASSERT(error);
+
+  delete error;
+}
+
+static int32_t detail_get_port_number(const char* port, error_s* error) {
+  SWEEP_ASSERT(port);
+  SWEEP_ASSERT(error);
+
+  if (strlen(port) <= 3) {
```

Comment:
I had originally been saving the end value in case we wanted to add a check for bad inputs with spaces, ie: "COM5 6". But I think what we have is sufficient. I switched it to use `nullptr`.